### ISSUE TITLE
Conflicting advice about quoting variables

### ISSUE THREAD

**Issue by bje-** [koalaman/shellcheck#1218]
#### For bugs

- Rule SC2076/SC2086
- Shellcheck version 0.4.7
- [X] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash

re=$1
echo $re
[[ foo =~ "$re" ]] && echo y
```

#### Here's what shellcheck currently says:

```
Line 4:
echo $re
     ^-- SC2086: Double quote to prevent globbing and word splitting.
 
Line 5:
[[ foo =~ "$re" ]] && echo y
          ^-- SC2076: Don't quote rhs of =~, it'll match literally rather than as a regex.
```

#### Here's what I wanted or expected to see:

I don't understand why I should not quote the variable containing the regular expression to protect against the variable being unset. If `$re` is unset, then the shell will complain that there is no right-hand argument to the `=~` binary operator. Why isn't the advice consistent between echoing `$re` and using it with a regex operator?

**Comment by jleffler** [comment]
See the Bash manual on the [`[[`](https://www.gnu.org/software/bash/manual/bash.html#index-_005b_005b) operator.  If the RHS of `=~` is all quoted, it isn't treated as a regular expression.  It's weird, but it is the way Bash is defined to work — for worse or worse.

**Comment by bje-** [comment]
I see. Thanks!
